woocommerce/woocommerce
# Template API: Expose template block id and order to client

**Please don't run away... I saw you look at the file change count and the number of lines changed! 🙀 I assure you, the actual changes in this PR are pretty small and confined to the new product editor. 😇 These changes are easy to test and verify.**

The changes in this PR break down as follows:

- **source changes, core: 217 lines**
- source changes, replacing `initBlock` and `useBlockProps` in `@woocommerce/product-editor` for all blocks: 313 lines
- test changes (mostly test data changes): 203 lines
- documentation: 70 lines
- changelogs: 21 lines
- config files (mostly boilerplate; because of new `@woocommerce/block-templates` project): 195 lines
- **lock files (because of new project): 2,799 lines**


### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR exposes the template block ID and block order to the client (JS) code, to enable easier development and debugging (since these values are needed when using the template API).

This PR implements the acceptance criteria of #40228:

- PHP: `AbstractBlock::get_formatted_template()` and `BlockContainerTrait::get_formatted_template()` includes the following attributes in its output:
    - `_templateBlockId`
    - `_templateBlockOrder`
- JS: A `registerWooBlockType` method that registers a block and auto-adds the following attributes to it's metadata:
    - `_templateBlockId`
    - `_templateBlockOrder`
- JS: A `useWooBlockProps` hook that takes the block's attributes as a param and returns the props from the core `useBlockProps` plus the following props:
    - `data-template-block-id`
    - `data-template-block-order`
- JS: All blocks in `@woocommerce/product-editor` updated to use `useWooBlockProps` and `registerWooBlockType`, which results in the following attributes on their DOM elements when rendered:
    - `data-template-block-id`
    - `data-template-block-order`

Closes #40228.



### How to test the changes in this Pull Request:



Automated unit tests (run locally or just verify they pass in GH on this PR):

1. PHP unit tests:

```
pnpm --filter=woocommerce env:test
pnpm test:unit:env --filter 'Automattic\\WooCommerce\\Tests\\Internal\\Admin\\BlockTemplates'
```

2. JS unit tests:

```
pnpm run test --filter=@woocommerce/block-templates
```

Manual testing:

1. Enable the **Product Block Editor** under **WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**.
2. Go to **Products** > **Add New**
3. Verify each block has the `data-template-block-id` and `data-template-block-id` attributes. You can do this one of two ways:
    - The hard way: view the Elements tab of the browser developer tools and look for these DOM attributes
    - The easy way: Install the [WooCommerce Block Template Inspector plugin](https://github.com/woocommerce/woocommerce-block-template-inspector) and focus (click) on each field.
 



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
